diag_manager/fms_diag_buffer.F90

```diff
@@ -0,0 +1,1546 @@
+                                                  & buff_sizes(5)))
+      allocate(integer(kind=i4_kind) :: this%counter(buff_sizes(1),buff_sizes(2),buff_sizes(3),buff_sizes(4), &
+                                                   & buff_sizes(5)))
+      allocate(integer(kind=i4_kind) :: this%count_0d(n_samples))
+      this%counter = 0
+      this%buffer_type = i4
+    type is (integer(kind=i8_kind))
+      allocate(integer(kind=i8_kind) :: this%buffer(buff_sizes(1),buff_sizes(2),buff_sizes(3),buff_sizes(4), &
+                                                  & buff_sizes(5)))
+      allocate(integer(kind=i8_kind) :: this%counter(buff_sizes(1),buff_sizes(2),buff_sizes(3),buff_sizes(4), &
+                                                  & buff_sizes(5)))
+      allocate(integer(kind=i8_kind) :: this%count_0d(n_samples))
+      this%counter = 0
+      this%buffer_type = i8
+    type is (real(kind=r4_kind))
+      allocate(real(kind=r4_kind) :: this%buffer(buff_sizes(1),buff_sizes(2),buff_sizes(3),buff_sizes(4),  &
+                                               & buff_sizes(5)))
+      allocate(real(kind=r4_kind) :: this%counter(buff_sizes(1),buff_sizes(2),buff_sizes(3),buff_sizes(4), &
+                                                & buff_sizes(5)))
+      allocate(real(kind=r4_kind) :: this%count_0d(n_samples))
+      this%counter = 0
+      this%buffer_type = r4
+    type is (real(kind=r8_kind))
+      allocate(real(kind=r8_kind) :: this%buffer(buff_sizes(1),buff_sizes(2),buff_sizes(3),buff_sizes(4), &
+                                               & buff_sizes(5)))
+      allocate(real(kind=r8_kind) :: this%counter(buff_sizes(1),buff_sizes(2),buff_sizes(3),buff_sizes(4), &
+                                                & buff_sizes(5)))
+      allocate(real(kind=r8_kind) :: this%count_0d(n_samples))
+      this%counter = 0
+      this%buffer_type = r8
+    class default
+       call mpp_error("allocate_buffer_5d", &
+           "The buff_type value passed to allocate a buffer is not a r8, r4, i8, or i4" // &
+           "for field:" // field_name_str, FATAL)
+  end select
+  allocate(this%num_elements(n_samples))
+  this%num_elements = 0
+  this%count_0d   = 0
+  allocate(this%buffer_dims(5))
+  this%buffer_dims(1) = buff_sizes(1)
+  this%buffer_dims(2) = buff_sizes(2)
+  this%buffer_dims(3) = buff_sizes(3)
+  this%buffer_dims(4) = buff_sizes(4)
+  this%buffer_dims(5) = buff_sizes(5)
+end subroutine allocate_buffer_5d
+
+!> Get routine for scalar buffers
+!! Sets the buff_out argument to the integer or real value currently stored in the buffer
+subroutine get_0d (this, buff_out, field_name)
+  class(buffer0d_type), intent(in) :: this !< 0d allocated buffer object
+  class(*), allocatable, intent(out)  :: buff_out !< output of copied buffer data
+  character(len=*), optional, intent(in) :: field_name !< optional field name for error output
```

Comment:
This should be required, but it's ok i guess.

Reply:
I made them required